torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,306 @@
+# method, and ZeRO requires a functional optimizer to overlap with DDP
+# Passing a `None` instead of an actual gradient indicates to the optimizer
+# to not update the corresponding parameter
+_NO_PARAM_UPDATE = None
+
+
+def hook_with_zero_step(
+    hook: Callable[[Any, dist.GradBucket], torch.futures.Future],
+    ddp: DistributedDataParallel,
+    zero: ZeroRedundancyOptimizer,
+) -> Callable[[Any, dist.GradBucket], torch.futures.Future]:
+    r"""
+    Modifies the given ``hook`` to overlap the :class:`ZeroRedundancyOptimizer`
+    optimizer step with the :class:`DistributedDataParallel` backward pass,
+    where the optimizer step computation begins after the last gradient bucket
+    computation has finished.
+
+    This approach overlaps the optimizer computation and communication with the
+    backward communication. In particular, the backward computation proceeds
+    contiguously, and the optimizer computation follows, overlapping with
+    outstanding backward communication (i.e. all-reduces) and possibly other
+    optimizer communication (i.e. broadcasts).
+
+    This approach may be preferred over :meth:`hook_with_zero_step_interleaved`
+    if communication is relatively slow compared to computation.
+
+    Arguments:
+        hook (Any * dist.GradBucket -> torch.futures.Future): the hook to
+            modify.
+        ddp (DistributedDataParallel): the :class:`DistributedDataParallel`
+            instance to use.
+        zero (ZeroRedundancyOptimizer): the :class:`ZeroRedundancyOptimizer`
+            instance to use.
+
+    Returns:
+        The modified hook.
+
+    Raises:
+        ValueError: if ``zero`` was constructed with ``overlap_with_ddp=False``.
+    """
+    if not zero._overlap_with_ddp:
+        raise ValueError(
+            "ZeroRedundancyOptimizer must be constructed with "
+            "`overlap_with_ddp=True` to use this hook properly"
+        )
+
+    def hook_with_zero_fn(
+        state,
+        bucket: dist.GradBucket,
+    ) -> torch.futures.Future:
+        r"""
+        Returns a :class:`Future` that gives a gradient bucket tensor and
+        performs the equivalent of a :class:`ZeroRedundancyOptimizer`
+        :meth:`step` if ``bucket`` is the last gradient bucket.
+
+        The function performs additional computation on the iteration that
+        the :class:`DistributedDataParallel` buckets are rebuilt to collect
+        information used to implement the modified hook.
+
+        Arguments:
+            state: any state for the hook.
+            bucket (dist.GradBucket): the :class:`DistributedDataParallel`
+                gradient bucket.
+        """
+        fut = hook(state, bucket)
+        overlap_info = zero._overlap_info
+        bucket_index = bucket.get_index()
+
+        # Proceed as normal until the DDP buckets have been rebuilt
+        if not ddp._has_rebuilt_buckets:
+            assert overlap_info.status == _OverlapStatus.UNINITIALIZED
```

Comment:
This means there is no optimizer step in the first 2 iterations and gradients are discarded? Do we need to mention this in the doc, or do you plan to change this in followup PRs?

Reply:
I added warnings to mention this in both hook functions and `ZeroRedundancyOptimizer`.

The motivation for this is so that we can avoid initializing a non-functional local optimizer and then re-initializing a functional optimizer soon thereafter. Even if we did that, we would have to transfer optimizer state correctly. 
We do not know until DDP buckets are built which parameters a given rank is assigned to update. I have not yet come up with a clean solution that allows not throwing away the first two inputs.

Reply:
Small note: The first **three** inputs are unused if `static_graph=True` for the DDP model.

Reply:
(can be done in followup PRs)

> We do not know until DDP buckets are built which parameters a given rank is assigned to update. I have not yet come up with a clean solution that allows not throwing away the first two inputs.

Since ZeRO's overlap states knows whether the buckets are rebuilt or not, is it possible that, in the first 2-3 iterations, `ZeroRedundancyOptimizer.step()` still performs comp + comm the same way as no-overlap cases? And only after the buckets are rebuilt, `ZeroRedundancyOptimizer.step()` becomes no-op?

Reply:
Yes, we can do that. Originally, I was averse to the idea of re-initializing the local optimizer, but upon further thought, providing an equivalent functionality to the user is definitely more important. 

What we can do is assume `parameter_as_bucket_view=False` behavior until buckets are rebuilt. This means that the parameters will be partitioned according to the normal sorted-greedy algorithm and a local optimizer will be initialized as normal at construction time. Then, when buckets are rebuilt, the parameter partitioning will change and the local optimizer will be reinitialized. I can do this in a follow-up PR.